### PR TITLE
[PGNCCL] Fix behavior of destroy_process_group

### PR DESCRIPTION
**Pull request by kwen2501**
Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* #141511
* __->__ #141510

Today `destroy_process_group()` is implemented via `ncclCommAbort`. 
When user call it in CPU, risk is that a healthy NCCL kernel gets preempted, which causes data corruption.

Instead of aborting kernels, we should flush collectives in `destroy_process_group`, i.e. let them complete normally, before we tear down resources. 

This PR implements such "flushing" behavior using `ncclCommFinalize`, then reclaims resources via `ncclCommDestroy`.

Expected behaviors:
For a bad program, a hang is expected at `destroy_process_group()`. If the PG uses non-blocking communicators, such hang is recoverable, because we attaches a timeout to the flush behavior.

cc @H-Huang @awgu @wanchaol @fegin @fduwjj @wz337 @wconstab @d4l3k @c-p-i-o

### CODE REVIEW

File: torch/csrc/distributed/c10d/ProcessGroupNCCL.cpp [hunk 1]
```diff
@@ -1419,13 +1419,55 @@ void ProcessGroupNCCL::abort() {
   monitorWakeUpCV_.notify_one();
 }
 
+// Difference between `abort()` and `shutdown()`:
+// 1. `abort()` will signal communicators to terminate all NCCL kernels
+// immediately.
+// 2. `shutdown()` will wait for all NCCL kernels to finish before destroying
```
Comment (fduwjj) [inline torch/csrc/distributed/c10d/ProcessGroupNCCL.cpp:1425]: Is `shutdown` blocking?

**Review by fduwjj — COMMENTED** [review]
Do you want to add a unit test for this?

File: torch/csrc/distributed/c10d/ProcessGroupNCCL.cpp (hunk 1 above)
Comment (kwen2501) [reply torch/csrc/distributed/c10d/ProcessGroupNCCL.cpp:1425]: Yes, it is blocking by purpose.

**Comment by kwen2501** [comment]
> Do you want to add a unit test for this?

`test_c10d_nccl.py` has ~ 50 calls of `destroy_process_group`. I think we can rely on them to test this change.

**Review by wconstab — APPROVED** [review]

**Pull request closed without merging**